lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -93,6 +93,27 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
 // Binary Operators
 //===----------------------------------------------------------------------===//
 
+OpFoldResult AddPrimOp::fold(ArrayRef<Attribute> operands) {
+  /// Any folding here requires a bitwidth extension.
+
+  /// If both operands are constant, and < 64 bits, then perform constant
+  /// folding. Cannot use constFoldBinaryOp, since the width of the constant is
+  /// different from operands.
+  if (operands[0] && operands[1] &&
+      lhs().getType().cast<FIRRTLType>().getBitWidthOrSentinel() < 64) {
+    auto op1 = operands[0].dyn_cast<IntegerAttr>();
+    auto op2 = operands[1].dyn_cast<IntegerAttr>();
+    if (op1 && op2) {
+      APInt sumAPInt = (op1.getValue() + op2.getValue());
+      int64_t sumVal = sumAPInt.isSignBitSet() ? sumAPInt.getSExtValue()
+                                               : sumAPInt.getZExtValue();
```

Comment:
Plz be careful to follow the implemention that LowerToRTL is using, this isn't correct.

In the case of add, we need to extend the LHS and RHS to the width of the destination (sign or zero extending depending on the type) and then do an add.  I'd recommend defining a helper function for this since the other constant folders will want to do a similar thing, e.g.:

`auto op1 = extendConstantToType(operands[0], getOperand(0), getType());` // returns an APInt

You need to pass in `getOperand(0)` because the sign/zero extension follows the sign of the operand, not the sign of the destination type.

Alternatively we could make sure that the attributes in FIRRTL::ConstantOp always have the correct sign.  If that were the case then we wouldn't need to pass in the type.  It looks like this could be true, but `verifyConstantOp` isn't verifying it, so it might not be true currently.

With this helper function then the body of this would just be:

```
auto op1 = extendConstantToType(operands[0], getOperand(0), getType());
auto op2 = extendConstantToType(operands[1], getOperand(1), getType());
return getIntAttr(op1+op2, getContext());
```

which also handles arbitrary width integers.